lib/iris/fileformats/_ugrid_cf_reader.py

```diff
@@ -0,0 +1,182 @@
+            mesh_var = dataset.variables[mesh.mesh_name]
+            if mesh.faces is not None:
+                # Work out name of faces dimension and record it.
+                faces_varname = mesh_var.face_node_connectivity
+                faces_var = dataset.variables[faces_varname]
+                if "face_dimension" in faces_var.ncattrs():
+                    faces_dim_index = faces_var.getncattr("face_dimension")
+                else:
+                    faces_dim_index = 0
+                faces_dim_name = faces_var.dimensions[faces_dim_index]
+                meshdims_map[faces_dim_name] = (mesh, "face")
+            if mesh.edges is not None:
+                # Work out name of edges dimension and record it.
+                edges_varname = mesh_var.edge_node_connectivity
+                edges_var = dataset.variables[edges_varname]
+                if "edge_dimension" in edges_var.ncattrs():
+                    edges_dim_index = faces_var.getncattr("face_dimension")
+                else:
+                    edges_dim_index = 0
+                edges_dim_name = edges_var.dimensions[edges_dim_index]
+                meshdims_map[edges_dim_name] = (mesh, "edge")
+            if mesh.nodes is not None:
+                # Work out name of nodes dimension and record it.
+                nodes_varname = mesh_var.node_coordinates.split()[0]
+                nodes_var = dataset.variables[nodes_varname]
+                nodes_dim_name = nodes_var.dimensions[0]
+                meshdims_map[nodes_dim_name] = (mesh, "node")
+        self.meshdims_map = meshdims_map
+
+        # Create a CFReader object which skips the UGRID-related variables.
+        kwargs["exclude_var_names"] = exclude_vars
+        self.cfreader = CFReader(self.dataset, *args, **kwargs)
+
+    def complete_ugrid_cube(self, cube):
+        """
+        Add the ".ugrid" property to a cube loaded with the `self.cfreader`.
+
+        We identify the unstructured-grid dimension of the cube (if any), and
+        attach a suitable CubeUgrid object, linking the cube mesh dimension to
+        an element-type (aka "mesh_location") of a mesh.
+
+        """
+        # Set a 'cube.ugrid' property.
+        data_var = self.dataset.variables[cube.var_name]
+        meshes_info = [
+            (i_dim, self.meshdims_map.get(dim_name))
+            for i_dim, dim_name in enumerate(data_var.dimensions)
+            if dim_name in self.meshdims_map
+        ]
+        if len(meshes_info) > 1:
+            msg = "Cube maps more than one mesh dimension: {}"
+            raise ValueError(msg.format(meshes_info))
+        if meshes_info:
+            i_dim, (mesh, mesh_location) = meshes_info[0]
+            cube.ugrid = CubeUgrid(
+                cube_dim=i_dim, grid=mesh, mesh_location=mesh_location
+            )
+        else:
+            # Add an empty 'cube.ugrid' to all cubes otherwise.
+            cube.ugrid = None
+        return
+
+    def __del__(self):
```

Comment:
NOTE: 
This class has taken over the responsibility of doing this, from the `iris.fileformats.cf.CFReader` class.
So `CFReader.__init__` is now passed a open dataset object instead of a filename.
I think that's o.k. as CFReader is not used anywhere else.

The point is that the UGridCFReader holds the reference to the open nc.Dataset, so we can do extra anaylsis based on the file content withou re-opening it.  The same open Dataset is now used by this code, the gridded.pyugrid calls, and the CFReader instance.